stability-tests/daa/daa_test.go

```diff
@@ -158,154 +163,186 @@ func measureMachineHashNanoseconds(t *testing.T) int64 {
 
-		_, err = rpcClient.SubmitBlock(templateBlock)
-		if err != nil {
-			t.Fatalf("SubmitBlock: %s", err)
+		// Exit early if the test is finished
+		if *isFinished {
+			return
 		}
+
+		submitMinedBlock(t, rpcClient, templateBlock)
 	})
 
 	averageBlocksPerSecond := calculateAverageDuration(miningDurations).Seconds()
 	expectedAverageBlocksPerSecond := float64(1)
 	deviation := math.Abs(expectedAverageBlocksPerSecond - averageBlocksPerSecond)
 	if deviation > blockRateDeviationThreshold {
-		t.Fatalf("Block rate deviation %f is higher than threshold %f. Want: %f, got: %f",
+		t.Errorf("Block rate deviation %f is higher than threshold %f. Want: %f, got: %f",
 			deviation, blockRateDeviationThreshold, expectedAverageBlocksPerSecond, averageBlocksPerSecond)
 	}
 }
 
-func hashNanosecondsToHashesPerSecond(hashNanoseconds int64) int64 {
-	return time.Second.Nanoseconds() / hashNanoseconds
+func fetchBlockForMining(t *testing.T, rpcClient *rpcclient.RPCClient) *externalapi.DomainBlock {
+	getBlockTemplateResponse, err := rpcClient.GetBlockTemplate(miningAddress)
+	if err != nil {
+		t.Fatalf("GetBlockTemplate: %s", err)
+	}
+	templateBlock, err := appmessage.RPCBlockToDomainBlock(getBlockTemplateResponse.Block)
+	if err != nil {
+		t.Fatalf("RPCBlockToDomainBlock: %s", err)
+	}
+	return templateBlock
 }
 
-func runForDuration(duration time.Duration, runFunction func()) {
-	isFinished := false
-	go func() {
-		for !isFinished {
-			runFunction()
+func tryNonceForMiningAndIncrementNonce(headerForMining externalapi.MutableBlockHeader, nonce *uint64,
+	targetDifficulty *big.Int, templateBlock *externalapi.DomainBlock) bool {
+
+	headerForMining.SetNonce(*nonce)
+	if pow.CheckProofOfWorkWithTarget(headerForMining, targetDifficulty) {
+		templateBlock.Header = headerForMining.ToImmutable()
+		return true
+	}
+
+	*nonce++
+	return false
+}
+
+func waitUntilTargetHashDurationHadElapsed(startTime time.Time, hashStartTime time.Time,
+	targetHashNanosecondsFunction func(totalElapsedDuration time.Duration) int64) {
+
+	// Yielding a thread in Go takes up to a few milliseconds whereas hashing once
+	// takes a few hundred nanoseconds, so we spin in place instead of e.g. calling time.Sleep()
+	for {
+		targetHashNanoseconds := targetHashNanosecondsFunction(time.Since(startTime))
+		hashElapsedDurationNanoseconds := time.Since(hashStartTime).Nanoseconds()
+		if hashElapsedDurationNanoseconds >= targetHashNanoseconds {
+			break
 		}
-	}()
-	time.Sleep(duration)
-	isFinished = true
+	}
+}
+
+func appendHashDuration(hashDurations []time.Duration, hashDuration time.Duration) []time.Duration {
+	hashDurations = append(hashDurations, hashDuration)
+	if len(hashDurations) > averageHashRateSampleSize {
+		hashDurations = hashDurations[1:]
+	}
+	return hashDurations
 }
 
-func runKaspad(t *testing.T) func() {
-	dataDir, err := common.TempDir("kaspad-daa-test")
+func appendMiningDuration(miningDurations []time.Duration, miningDuration time.Duration) []time.Duration {
```

Comment:
Consider combining appendHashDuration and appendMiningDuration, they are functionally the same.

Maybe create a type `durationsSlice` or whatever, that has a `maxItems` and an `append` and `average` methods.